fs/inode/fs_files.c

```diff
@@ -304,54 +433,35 @@ static int nx_dup3_from_tcb(FAR struct tcb_s *tcb, int fd1, int fd2,
 
   if (fd2 >= count)
     {
-      ret = files_extend(list, fd2 / CONFIG_NFILE_DESCRIPTORS_PER_BLOCK + 1);
+      ret = fdlist_extend(list,
+                          fd2 / CONFIG_NFILE_DESCRIPTORS_PER_BLOCK + 1);
       if (ret < 0)
         {
           return ret;
         }
     }
 
-  filep1 = files_fget(list, fd1);
+  /* Get file behind fd1, and keep the reference */
+
+  filep1 = fget_by_fd(list, fd1, &fdflags);
   if (filep1 == NULL)
     {
       return -EBADF;
     }
 
-  filep = files_fget_by_index(list,
-                              fd2 / CONFIG_NFILE_DESCRIPTORS_PER_BLOCK,
-                              fd2 % CONFIG_NFILE_DESCRIPTORS_PER_BLOCK,
-                              &new);
+  /* Atomically swap the file pointer behind fd2 */
 
-#ifdef CONFIG_FDSAN
-  f_tag_fdsan = filep->f_tag_fdsan;
-#endif
-
-#ifdef CONFIG_FDCHECK
-  f_tag_fdcheck = filep->f_tag_fdcheck;
-#endif
+  fdflags.f_cloexec = flags == FD_CLOEXEC ? true : false;
+  filep2 = fdlist_finstall(list, fd2, filep1, &fdflags);
 
-  /* Perform the dup3 operation */
+  /* Close the old file, if any */
 
-  ret = file_dup3(filep1, filep, flags);
-  fs_putfilep(filep1);
-  fs_putfilep(filep);
-  if (ret < 0)
+  if (filep2 != NULL)
     {
-      if (new)
-        {
-          fs_putfilep(filep);
-        }
-
-      return ret;
+      fs_putfilep(filep2);
```

Comment:
Line 455 removes the file from the fdlist, and points filep1 to fd2. Then filep2 is closed here. This works exactly like POSIX requires.

Reply:
See it and thanks